codalab/worker/bundle_manager.py

```diff
@@ -0,0 +1,389 @@
+            parent_uuids = set(dep.parent_uuid for dep in bundle.dependencies)
+            parents = self._model.batch_get_bundles(uuid=parent_uuids)
+            parent_dict = {parent.uuid: parent for parent in parents}
+
+            bundle.install_dependencies(self._bundle_store, parent_dict, path, copy=True)
+
+            self._upload_manager.update_metadata_and_save(bundle, new_bundle=False)
+            logger.info('Finished making bundle %s', bundle.uuid)
+            self._model.update_bundle(bundle, {'state': State.READY})
+        except Exception as e:
+            logger.info('Failing bundle %s: %s', bundle.uuid, str(e))
+            self._model.update_bundle(
+                bundle, {'state': State.FAILED,
+                         'metadata': {'failure_message': str(e)}})
+        finally:
+            with self._make_uuids_lock:
+                self._make_uuids.remove(bundle.uuid)
+
+    def _cleanup_dead_workers(self, workers, callback=None):
+        """
+        Clean-up workers that we haven't heard from for more than 5 minutes.
+        Such workers probably died without checking out properly.
+        """
+        for worker in workers.workers():
+            if datetime.datetime.now() - worker['checkin_time'] > datetime.timedelta(minutes=5):
+                logger.info('Cleaning up dead worker (%s, %s)', worker['user_id'], worker['worker_id'])
+                self._worker_model.worker_cleanup(worker['user_id'], worker['worker_id'])
+                workers.remove(worker)
+                if callback is not None:
+                    callback(worker)
+
+    def _restage_stuck_starting_bundles(self, workers):
+        """
+        Moves bundles that got stuck in the STARTING state back to the STAGED
+        state so that they can be scheduled to run again.
+        """
+        for bundle in self._model.batch_get_bundles(state=State.STARTING, bundle_type='run'):
+            if (not workers.is_running(bundle.uuid) or  # Dead worker.
+                time.time() - bundle.metadata.last_updated > 5 * 60):  # Run message went missing.
+                logger.info('Re-staging run bundle %s', bundle.uuid)
+                if self._model.restage_bundle(bundle):
+                    workers.restage(bundle.uuid)
+
+    def _fail_stuck_running_bundles(self, workers):
+        """
+        Fails bundles that got stuck in the RUNNING state.
+        """
+        for bundle in self._model.batch_get_bundles(state=State.RUNNING, bundle_type='run'):
+            if (not workers.is_running(bundle.uuid) or  # Dead worker.
+                time.time() - bundle.metadata.last_updated > 60 * 60):  # Shouldn't really happen, but let's be safe.
+                failure_message = 'Worker died'
+                logger.info('Failing bundle %s: %s', bundle.uuid, failure_message)
+                self._model.finalize_bundle(bundle, -1, exitcode=None, failure_message=failure_message)
+                workers.unqueue(bundle.uuid)
```

Comment:
Oops, that should be restage...

On Wed, May 18, 2016 at 9:12 PM, Stephen Koo <notifications@github.com>
wrote:

> In codalab/worker/bundle_manager.py
> <https://github.com/codalab/codalab-cli/pull/428#discussion_r63820679>:
>
> > +                time.time() - bundle.metadata.last_updated > 5 * 60):  # Run message went missing.
> > +                logger.info('Re-staging run bundle %s', bundle.uuid)
> > +                if self._model.restage_bundle(bundle):
> > +                    workers.restage(bundle.uuid)
> > +
> > +    def _fail_stuck_running_bundles(self, workers):
> > +        """
> > +        Fails bundles that got stuck in the RUNNING state.
> > +        """
> > +        for bundle in self._model.batch_get_bundles(state=State.RUNNING, bundle_type='run'):
> > +            if (not workers.is_running(bundle.uuid) or  # Dead worker.
> > +                time.time() - bundle.metadata.last_updated > 60 * 60):  # Shouldn't really happen, but let's be safe.
> > +                failure_message = 'Worker died'
> > +                logger.info('Failing bundle %s: %s', bundle.uuid, failure_message)
> > +                self._model.finalize_bundle(bundle, -1, exitcode=None, failure_message=failure_message)
> > +                workers.unqueue(bundle.uuid)
>
> @klopyrev <https://github.com/klopyrev> Where is this method defined? I
> can't seem to find it.
>
> I'm hitting the following error when testing the worker system:
>
> Traceback (most recent call last):
>   File "/Users/sckoo/Developer/codalab-cli/codalab/worker/bundle_manager.py", line 76, in run
>     self._run_iteration()
>   File "/Users/sckoo/Developer/codalab-cli/codalab/worker/bundle_manager.py", line 96, in _run_iteration
>     self._schedule_run_bundles()
>   File "/Users/sckoo/Developer/codalab-cli/codalab/worker/default_bundle_manager.py", line 30, in _schedule_run_bundles
>     self._fail_stuck_running_bundles(workers)
>   File "/Users/sckoo/Developer/codalab-cli/codalab/worker/bundle_manager.py", line 239, in _fail_stuck_running_bundles
>     workers.unqueue(bundle.uuid)
> AttributeError: 'WorkerInfoAccessor' object has no attribute 'unqueue'
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/codalab/codalab-cli/pull/428/files/4a9bd3548d5d52c0bba561495c0d5e03aef63e8b#r63820679>
>